test_spack.py

```diff
@@ -320,8 +326,20 @@ def Self_and_up(origin: set, DAG: map) -> set:
             return reachers
 
 
-def Has_cycle(directed_graph: map) -> bool:
-    return False  # TODO: Implement!
+def Has_cycle(directed_graph: map, visited=None, vertex=None) -> bool:
```

Comment:
I don't understand the meaning of this function. What cycle do you check?
If there are circular dependencies?

Reply:
This function tests if the directed graph has a cycle. If I call the function on the dependency graph, it tells me if there are circular dependencies.